Replace the InternalChangeReplicas raft command with a distributed txn.

bdarnell opened the pull request:
This allows us to update the addressing records as well, but I'm leaving
that for a subsequent commit.

multiraft.ChangeGroupMembership can now smuggle a command to be executed
atomically with the membership change; we do this to attach the
membership change to the EndTransaction call.

storage/raft_test.go has been moved to client_test.go and
client_raft_test.go now that it needs to use distributed transactions.

bdarnell committed:
Replace the InternalChangeReplicas raft command with a distributed txn.

This allows us to update the addressing records as well, but I'm leaving
that for a subsequent commit.

multiraft.ChangeGroupMembership can now smuggle a command to be executed
atomically with the membership change; we do this to attach the
membership change to the EndTransaction call.

storage/raft_test.go has been moved to client_test.go and
client_raft_test.go now that it needs to use distributed transactions.

spencerkimball commented on proto/api.proto:
```diff
@@ -234,6 +234,7 @@ message EndTransactionRequest {
   // public-facing KV API.
   optional SplitTrigger split_trigger = 3;
   optional MergeTrigger merge_trigger = 4;
+  optional ChangeReplicasTrigger change_replicas_trigger = 5;
```
You're going to need to update kv/db.go and kv/db_test.go to properly ignore this value if set through the public-facing API. We probably should make the trigger a proto buf message itself and move its definition to internal.proto so we don't have to keep adding these cases. Might as well add a TODO about that.

bdarnell replied:
Done.

spencerkimball commented on storage/range.go:
```diff
@@ -1759,27 +1760,64 @@ func (r *Range) AdminMerge(args *proto.AdminMergeRequest, reply *proto.AdminMerg
+			break
+		}
+	}
+	if changeType == proto.ADD_REPLICA {
```
I think this algorithm will work for adding and removing. But keep in mind that we have an issue here because we need to be precise about which replica is being replaced. Our zone configs specify attributes for replica indexes. We can't just let the replicas get jumbled up. If we always add a new replica in order to replace an existing, this feels like it will work if the order is always Add, then Remove, with no chance for another replica to be added first.

bdarnell replied:
Really? Tying configs to replica order seems fragile; I assumed that the zone configs specified attributes of the set as a whole. (my previous commit on this topic reordered Desc.Replicas arbitrarily on any change). This algorithm can work for add-then-remove (assuming we maintain some sort of lock so there are no other changes in between), but it breaks down for remove-then-add, and if we're changing from 3 to 5 replicas or vice versa it gets complicated to maintain the desired ordering without causing unnecessary churn at the raft level (if we add the replica at index 5 before the one at index 4 do we need to create a gap in the list?)

tschottdorf replied:
I wasn't aware of that either and doing a quick search I didn't find anything that obviously relied on this. Can't we get away with treating `[]proto.Replica` as a set and adding something like a:

`func (zc *ZoneConfig) Check([]proto.Replica) (orderedSatisfiedReplicas []proto.Replica, diffWeNeedToFix something)`

whenever we need this. Is this relevant outside of the scanner queue which applies the Zone config (which I don't think exists yet)?

I agree with @bdarnell that relying on the order seems unexpected and tricky.

spencerkimball commented on storage/range.go:
```diff
@@ -1759,27 +1760,64 @@ func (r *Range) AdminMerge(args *proto.AdminMergeRequest, reply *proto.AdminMerg
 	}
+	err := r.rm.DB().RunTransaction(txnOpts, func(txn *client.KV) error {
+		// Important: the range descriptor must be the first thing touched in the transaction
+		// so the transaction record lives there.
```
s/the transaction record lives there/the transaction record is co-located with the range being modified/

bdarnell replied:
Done

spencerkimball commented on storage/range.go:
```diff
@@ -1759,27 +1760,64 @@ func (r *Range) AdminMerge(args *proto.AdminMergeRequest, reply *proto.AdminMerg
 
-	if err := engine.MVCCPutProto(batch, nil, engine.RangeDescriptorKey(r.Desc.StartKey), req.Header().Timestamp, nil, r.Desc); err != nil {
-		reply.SetGoError(util.Errorf("change replicas of range %d failed: %s", r.Desc.RaftID, err))
+		// End the transaction manually instead of letting RunTransaction
```
Add a TODO here to call UpdateRangeAddressing().

bdarnell replied:
Done

spencerkimball commented on multiraft/multiraft.go:
```diff
@@ -720,6 +722,7 @@ func (s *state) handleWriteResponse(response *writeResponse, readyGroups map[uin
 					NodeID:     NodeID(cc.NodeID),
 					ChangeType: cc.Type,
 					ConfState:  *cs,
+					Payload:    payload,
```
What I'm concerned about here is that the conf change is applied before we know whether or not the transaction will commit. What if the transaction fails to commit?

bdarnell replied:
I'm working on this part now: adding a callback to EventCommandCommitted so we don't call multiNode.ApplyConfChange until the application is ready (and we can abort the change on error). What's the best way to force the transaction to fail for testing?

bdarnell committed:
Reject ChangeReplicasTrigger sent by client.

Update comments.

spencerkimball commented:
> Really? Tying configs to replica order seems fragile; I assumed that the zone configs specified
> attributes of the set as a whole. (my previous commit on this topic reordered Desc.Replicas arbitrarily 
> on any change). This algorithm can work for add-then-remove (assuming we maintain some sort of lock 
> so there are no other changes in between), but it breaks down for remove-then-add, and if we're 
> changing from 3 to 5 replicas or vice versa it gets complicated to maintain the desired ordering without
> causing unnecessary churn at the raft level (if we add the replica at index 5 before the one at index 4 
> do we need to create a gap in the list?)

It simplifies things to deal with indexes but if it's really going to be a pain, I could make it more flexible. Also, what is your plan for node attributes here? They really should be part of the request so they get set in the replica.

bdarnell commented:
The attributes get propagated automatically when the updated Desc is written in the transaction. I guess each replica needs to re-read its Desc from the store after the transaction commits.

spencerkimball commented on storage/range.go:
```diff
@@ -940,6 +939,8 @@ func (r *Range) EndTransaction(batch engine.Engine, args *proto.EndTransactionRe
 		} else if args.MergeTrigger != nil {
 			reply.SetGoError(r.mergeTrigger(batch, args.MergeTrigger))
+		} else if args.ChangeReplicasTrigger != nil {
+			// ChangeReplicasTrigger is processed on input, not output.
```
Actually, I think here you'll want to actually call a changeReplicasTrigger method which could then set the updated replicas in the descriptor. You might want to modify ChangeReplicasTrigger protobuf to accept an UpdatedDesc for this purpose.

bdarnell replied:
Yeah, I agree we need to call changeReplicasTrigger to update r.Desc. I specifically want to avoid accepting an updatedDesc in ChangeReplicas, however, because I don't want to have to compare the new and existing versions to make sure there is exactly one difference.

spencerkimball replied:
Maybe factor out the code in ChangeReplicas which creates the updated desc and call it both from ChangeReplicas and from changeReplicasTrigger.

bdarnell replied:
Oh wait, I misread. I don't want ChangeReplicas() to take an updatedDesc, but adding it to ChangeReplicasTrigger is fine.

bdarnell committed:
Add a Callback to EventMembershipChangeCommitted.

The change is no longer applied to raft until this callback is called;
if it is passed an error the change is aborted. This ensures that the
raft state does not change unless the transaction is committed.

bdarnell committed:
Update the in-memory Range.Desc after a ChangeReplicas commit.

bdarnell commented:
New commits added to only apply the config change if the transaction commits (need some pointers on testing this) and to update the in-memory range descriptor after the commit.

spencerkimball commented:
LGTM

As to testing...hard to make the end transaction abort. You could start another transaction with max priority and write the range descriptor and leave the transaction open. That will cause the ChangeReplicas txn to backoff and retry. You could wait a moment and verify that raft hasn't taken the replica change to heart, then abort the original transaction and verify that raft does get the change. It all seems pretty clumsy though.

Maybe it would be better to introduce a unittest-only mechanism for returning an error from inside the ChangeReplicas txn. You'd then set that to true for this particular test, verify the error, but make sure that the replica change didn't stick within Raft.

bdarnell committed:
Add a (hacky) test for aborted ChangeReplicas transactions.

bdarnell committed:
Fix a race condition by moving the membership change callback to the
raft thread.

bdarnell committed:
Only pass replicas through ChangeReplicasTrigger, not a full Desc.

bdarnell committed:
Use a two-phase stopper in eventDemux.

Fixes a race condition.

bdarnell committed:
Re-submit pending proposals after a membership change is applied.

Fixes a non-deterministic test failure.

bdarnell commented:
Updated the branch with a new test and some other fixes. The test is ugly but works, and we can use this hook to simulate other kinds of failures in the future.

tschottdorf commented on multiraft/multiraft.go:
```diff
@@ -696,30 +703,49 @@ func (s *state) handleWriteResponse(response *writeResponse, readyGroups map[uin
+								s.multiNode.ApplyConfChange(groupID, cc)
+							} else {
+								log.Warningf("aborting configuration change: %s", err)
+								s.multiNode.ApplyConfChange(groupID,
```
why apply an empty conf change here?

bdarnell replied:
It's an undocumented special case in raft: only one pending configuration change is allowed at a time; an empty ConfChange aborts the pending change without doing anything.

tschottdorf replied:
[From my snippet above](https://github.com/cockroachdb/cockroach/pull/364/files#diff-dfd5fe6017011d7ffff21e9d95d739c3R175) it looks like Raft simply discards the new ConfChange rather than abort the previous one, what am I missing?

bdarnell replied:
That snippet says that *proposing* a new ConfChange while one is pending is a no-op. To clear the pendingConf flag and let a new config, a change must be applied: the original ConfChange to commit it or an empty one to abort. The internals of ApplyConfChange where this happens are at https://github.com/coreos/etcd/blob/c7146bd5f2c73716091262edc638401bb8229144/raft/node.go#L291

tschottdorf replied:
Gotcha. I checked `multinode.go:198` just to make sure it's there as well, but of course it is.

tschottdorf commented on proto/api.go:
```diff
@@ -87,28 +87,27 @@ func (s stringSet) keys() []string {
 
 // AllMethods specifies the complete set of methods.
 var AllMethods = stringSet{
```
Unrelated, but I don't see this used anywhere, and it should always be equal to the union of {Internal,Read,Write}Methods anyways.

tschottdorf commented on storage/range.go:
```diff
@@ -159,10 +168,11 @@ type RangeManager interface {
+	rm    RangeManager // Makes some store methods available
+	stats *rangeStats  // Range statistics
+	// 1 if a split, merge, or replica change is underway; updated atomically
+	metaLock int32
```
is there another safety layer for this on the level of Raft? What would happen if you trigger a split on the leader, elect a new leader and on it, trigger another split?

tschottdorf replied:
Ah, there's one.
```
if e.Type == pb.EntryConfChange {
  if r.pendingConf {
	m.Entries[i] = pb.Entry{Type: pb.EntryNormal}
  }
  r.pendingConf = true
}
```

bdarnell commented:
@spencerkimball Want to take a look at the test here or is it ready to go?

bdarnell committed:
Add a TODO to verify raft internal state

spencerkimball commented:
LGTM
